File: src/main/java/net/hectus/bb/turn/counter/CounterFilter.java
package net.hectus.bb.turn.counter;

import com.marcpg.libpg.lang.Translation;
import net.hectus.bb.turn.TurnData;
import net.kyori.adventure.text.Component;
import org.jetbrains.annotations.NotNull;

import java.util.Locale;
import java.util.function.Predicate;

public interface CounterFilter {
    boolean doCounter(@NotNull TurnData last);
    String text(Locale l);

    default Component line(Locale l) {
        return Component.text("   | " + text(l));
    }

    // Have to use this, because @FunctionalInterface obviously wouldn't work with two methods.
    static @NotNull CounterFilter of(Predicate<TurnData> doCounter, String name) {
        return new CounterFilter() {
            @Override
            public boolean doCounter(@NotNull TurnData last) {
                return doCounter.test(last);
            }

            @Override
            public String text(Locale l) {
                return Translation.string(l, "item-lore.counter." + name);
            }
        };
    }
}

File: src/main/kotlin/net/hectus/neobb/modes/turn/default_game/attribute/CounterFilter.kt
package net.hectus.neobb.modes.turn.default_game.attribute

import com.marcpg.libpg.lang.string
import com.marcpg.libpg.util.component
import net.hectus.neobb.modes.turn.TurnExec
import net.hectus.neobb.util.Constants
import net.hectus.neobb.util.noItalic
import net.kyori.adventure.text.Component
import java.util.*

interface CounterFilter {
    companion object {
        fun of(name: String, doCounter: (TurnExec<*>) -> Boolean) = object : CounterFilter {
            override fun doCounter(last: TurnExec<*>): Boolean = doCounter.invoke(last)
            override fun text(locale: Locale): String = locale.string("item-lore.counter.$name")
        }
    }

    fun doCounter(last: TurnExec<*>): Boolean
    fun text(locale: Locale): String

    fun line(locale: Locale): Component = component("${Constants.MINECRAFT_TAB_CHAR}| " + text(locale)).noItalic()
}
